composer/callbacks/checkpoint_saver.py

```diff
@@ -509,17 +580,33 @@ def _save_checkpoint(self, state: State, logger: Logger):
                 ).lstrip('/') + '.symlink'
 
                 # create and upload a symlink file
-                with tempfile.TemporaryDirectory() as tmpdir:
-                    symlink_filename = os.path.join(tmpdir, 'latest.symlink')
-                    # Sharded checkpoints for torch >2.0 use directories not files for load_paths
-                    if state.fsdp_sharded_state_dict_enabled:
-                        src_path = str(pathlib.Path(remote_file_name).parent)
+                symlink_filename = os.path.join(
+                    self.tmp_dir_for_symlink.name,
+                    f'latest.{self.symlink_count}.symlink',
+                )
+                # Sharded checkpoints for torch >2.0 use directories not files for load_paths
+                if state.fsdp_sharded_state_dict_enabled:
+                    src_path = str(pathlib.Path(remote_file_name).parent)
+                else:
+                    src_path = remote_file_name
+                log.debug(f'Creating symlink file {symlink_filename} -> {src_path}')
+                rank_saves_symlinks = dist.get_global_rank() == 0 or not state.fsdp_sharded_state_dict_enabled
+                if rank_saves_symlinks:
+                    self.rank_saves_remote_symlinks = True
+                    create_symlink_file(src_path, symlink_filename)
+                    if self.remote_uploader is not None:
+                        remote_checkpoint_file_names = []
+                        for file_names in all_remote_filenames:
+                            remote_checkpoint_file_names += file_names
```

Comment:
oh because `dist.all_gather_object` returns a list of objects. so `all_remote_filenames` is a list of list, `remote_checkpoint_file_names` is a flatten list.